### PR TITLE
refactor: split `MetalsLanguageServer` into language server and language service

### PR DESCRIPTION
**Pull request by kpodsiad**
Previously, in `MetalsLanguageServer` there were multiple services which initialization has been delayed (`= _`) because project workspace or client config was needed. This is not needed as former `MetalsLanguageServer` can be instantiated only after receiving `Initialize` request from the client:

```mermaid
sequenceDiagram
participant C as LSP client
participant S as LSP server

C->>S: Initialize request <br> (workspace + client config)
S->>C: Initiliazed response
Note over C,S: Client must not send any <br> request until receiving <br> Initialized response
C-->>S: Initialized notification
Note over C,S: The Initialized notification is send <br> before client can send any other <br> request or notification to the server
```

Based on that knowledge I extracted those parts as `LanguageServer` trait. Similarly, I extracted `TextDocumentService`, `WorkspaceService` and `MetalsService` into traits. Traits are needed to create default, dummy implementation. This is due to how [JsonDelegate](https://github.com/eclipse/lsp4j/blob/main/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/services/JsonDelegate.java) works - in runtime class is being processed and if method annotated with delegate returns null then no methods will be registered. In runtime, at the beginning we can't provide valid implementation of those services (because we delay init after `Initialize` request) - hence traits with default, throwing implementation. 
Why I didn't used [lsp4j's services](https://github.com/eclipse/lsp4j/tree/main/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/services) - well sometimes return types of methods differ and I didn't dive if there's difference in how lsp4j handles Unit vs Future[Unit]

todo: 
- [x] pass workspace and client config as constructor args in old `MetalsLanguageServer`, move initialization logic to the body
- [x] reduce mutability when possible (and easy)
- [x] rename old `MetalsLanguageServer` to e.g. `MetalsLanguageServer`. Any better ideas? I want to rename that just before merge to make conflicts easier to resolve
- [x] fix tests:
- [x] fix`StatusBarSuite` -  there was an issue with `ClientConfiguration`
- [x] fix `CallHierarchyLspSuite` - split tests to not call "initialize" twice
- [x] fix some issues with thread pools cancelation - add scaladoc to make intentions clearer
- [x] fix integration tests
- [x] investigate issues with RejectedExecutionException which are visible in logs. Although, as @tgodzik pointed, in tests server is being cancelled after (to be precise before in beforeEach) each test and thread pools are being shutdown after each suite. 
- [x] double check cancelation logic, make sure that there is no resource leak

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:104]: Or MetalsServerWorkspace ?

**Comment by kpodsiad** [comment]
Ok, CI is quite ok right now, macos is failing but I believe those are flaky tests. One thing which still worries me - there are `org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.nio.channels.AsynchronousCloseException` and `java.util.concurrent.RejectedExecutionException: Task Future() rejected from` exceptions in logs. They might be cause by my changes, but it seems that they're not affecting correctness of Metals.

**Comment by tgodzik** [comment]
> Ok, CI is quite ok right now, macos is failing but I believe those are flaky tests. One thing which still worries me - there are `org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.nio.channels.AsynchronousCloseException` and `java.util.concurrent.RejectedExecutionException: Task Future() rejected from` exceptions in logs. They might be cause by my changes, but it seems that they're not affecting correctness of Metals.

I think it's because we shutdown the server mutiple times in the tests and the errors/warnings are legit, but not an issue.

**Comment by kpodsiad** [comment]
I've squashed commits, only meaningful ones remained. PR is ready for review.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:226]: `() => workspace` can be changed to just `workspace`

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:290]: Extension method can operate on plain `InitializeParams` rather than Option of it

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:807]: `() => tables` => just `tables`

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:262]: Maybe it's worth to add logging form each branch, just in case?

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:270]: Used only in tests, I didn't want to alter tests more than it was needed to fix failing ones. Additional work can be done in follow up of this PR

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:96]: Again, excuse the naive question, but related to your comment above. In this situation for example we only use `set` and `get`, but are still using an `AtomicBoolean` over a `@volatile`. How come?

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:28]: Bear with me, I'll probably ask a bunch of these just t make sure I'm tracking, but in case the dummy on is the `new ScalaLspService {}` which will then later be replaced by the `MetalsLspService` as the _real_ one right?

File: metals/src/main/scala/scala/meta/metals/lsp/LanguageServer.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/metals/lsp/LanguageServer.scala:49]: Let's add some scaladocs here just to make it super clear that this is the Dummy instance that doesn't do anything.

File: metals/src/main/scala/scala/meta/metals/lsp/MetalsService.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/metals/lsp/MetalsService.scala:27]: Out of curiosity, do these even need a default implementation then? Couldn't they just all be abstract?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:103]: I think `MetalsLspService` is fine. Could we finish up the scaladocs here as well?

File: metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala:39]: Just to be sure, we did this before because we wanted to ensure we got the configured client, but in this scenario, we _always_ have the configured client right?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:226]: Just to be sure, this is because the workspace is already known right?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:290]: Are you sure? I didn't see extensions methods in `MetalsEnrichments` for this. Or do you mean we should switch that?

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:262]: Are you referring to the shutdown/exits? If so sure.

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:270]: I'd say create an issue right away for this then to outline what would need to be changed and to remove this after.

**Review by ckipp01 — COMMENTED** [review]
Really really fantastic refactoring @kpodsiad. I love seeing all of the setting to `_` removed. This really cleans some stuff up. I just mainly have a couple questions/nits.

Since quite a bit of stuff is changing at the moment, would you also be able to run through https://github.com/scalameta/metals/blob/main/architecture.md and maybe update / expand the relevant parts about the changes that were made here?

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:88]: Only saves a few keystrokes when getting and setting value -  `=` vs `get()/set()`. On the other hand I wrote 2 lines of comment to describe why volatile is good enough 😅 
Maybe it's simpler to just use `AtomicReference` and _not think of_ anything

File: metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala:39]: Exactly. Previously, language client was obtained and set after creating MetalsLanguageServer - https://github.com/scalameta/metals/blob/3cbb04ef03d2d390b228744aa5c5824aa6a8274e/metals/src/main/scala/scala/meta/metals/Main.scala#L68-L70.

Now, with MetalsLanguageServer being refactored and new MetalsLspService, the later one is created with workspace and proper client passed in so there is no need for `() => ` workarounds

Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/StatusBar.scala:39]: [Here](https://github.com/kpodsiad/metals/blob/908c0c1e199626abcf3375d96555772568e794ab/metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala#L176-L192), we're creating MetalsService

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:262]: yes, both cases (happy and error paths) for both shutdown and exit

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:290]: It is `XtensionClientCapabilities` which operates on `initializeParams: Option[l.InitializeParams]`

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:226]: yup

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:96]: good point, I'm not being consistent. As I wrote earlier, it's probably easier to switch to `Atomic*` and just forget about `@volatile`

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:28]: yes, you're correct.
The thing is, at runtime (I'm not sure at exactly which point though), probably betwenn those lines https://github.com/kpodsiad/metals/blob/908c0c1e199626abcf3375d96555772568e794ab/metals/src/main/scala/scala/meta/metals/Main.scala#L59-L70 lsp4j processes given server which was set in `setLocalService` method. If method with `JsonDelegate` annotation exists, that method is being called. If it returns null processing is aborted, if not them processing is continued on class which is being returned. 

The tricky part - now we delay creation of MetalsLspService so we cannot provide working instance from the very beginning. We can't return null because lsp4j processing won't process `ScalaLspService` methods and our MetalsLanguageServer will "see" only basic init/shutdown endpoints*

* - it is also possible that I misunderstood some parts of how lsp4j works, but after I reached current state I was happy enough to see things working.

Answering your questions in this thread so we can keep discussion in one place:

`ScalaLspService` isn't just a Dummy instance, it's an interface which describes endpoints. That interface is being later on implemented by MetalsLspService. What may be con fusing is fact, that `ScalaLspService` mixes two roles:
- it's interface
- it provides default implementation for methods so it is possible to do just `new ScalaLspService {}`

Default implementations are needed because of later point. Otherwise all of those methods had to be implemented

Are things more clear now? Maybe scaladocs needs some improvement?

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:290]: Yes, exactly, that's what I mean, it operates on an `Option`. Does your original comment mean you'd like to change this to operate directly on `l.InitializeParams` instead of the `Option`?

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:88]: ha, I don't have strong feelings on it, was just curious. Either way, I think consistency is important. So if we go with `@volatile`, then we should do it for all that we can. If not, then yea, let's just stick with the `AtomicReference`.

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:28]: Yea this helps, thanks.

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:60]: yes we can, but I'd delay it a bit. I think we can do much better job when it comes to configuration - https://github.com/scalameta/metals/issues/4789

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:28]: Should I modify scaladoc to make things crystal clear?

**Comment by kpodsiad** [comment]
@ckipp01 I think that I addressed your comments either directly or by creating follow up tasks and placing todos in code. I'd like to merge this PR as soon as it is possible to test it thoroughly before next release.

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:28]: I think more than a Scaladoc we should try and outline the entire thing in the design doc.

**Comment by ckipp01** [comment]
> @ckipp01 I think that I addressed your comments either directly or by creating follow up tasks and placing todos in code. I'd like to merge this PR as soon as it is possible to test it thoroughly before next release.

I think we want to release before we merge this no? It's been quite some time before we released, so I'd actually recommend trying to finish up https://github.com/scalameta/metals/pull/4740 as quickly as we can, cut a release, and then get this merged. I wouldn't try to merge this before that release since there is already a ton of stuff in there.

@tgodzik is there anything I can help with for the release or are there blockers yet?

**Comment by kpodsiad** [comment]
> I think we want to release before we merge this no? It's been quite some time before we released, so I'd actually recommend trying to finish up #4740 as quickly as we can, cut a release, and then get this merged. I wouldn't try to merge this before that release since there is already a ton of stuff in there.

Yes, that's the idea, to merge it after current release.

**Comment by kpodsiad** [comment]
I renamed old `MetalsLanguageServer` to `MetalsLanguageService`. I'll update `architecture.md` in separate PR.
@ckipp01 @tgodzik can I have an approval on this one? ;)

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:43]: This one should be fairly simple right? Are you sure you don't want to just tackle this right away? Or do you still want to merge and then send in a follow-up for this? The reason I ask is because you already dit this for `tables` in here. So its odd that you did one but not the others.

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:51]: 👍🏼 as long as you plan on following up with those other issues, then no worries. If you aren't going to get to them right away, then I'd say add them now.

**Review by ckipp01 — CHANGES_REQUESTED** [review]
The only blocker at this point for me is making sure that https://github.com/scalameta/metals/blob/main/architecture.md is updated. I do have a few small nits, but if you plan on following up with the other issues, then no worries.

- [ ] Update architecture.md

**Comment by ckipp01** [comment]
> I'll update architecture.md in separate PR.

pfff, I totally missed this in your last comment.  😬

**Review by ckipp01 — APPROVED** [review]
LGTM from me, but I think you should also wait for another ✅ from @tgodzik since this is a huge change.

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:51]: I had this in mind, but I also see that @tanishiking assigned @kasiaMarek in https://github.com/scalameta/metals/issues/4789. It seems that definitely someone will handle this ;)

File: metals/src/main/scala/scala/meta/internal/metals/Indexer.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/Indexer.scala:43]: One should be really careful when replacing `() =>`, in most cases this will be straighforward change, but sometimes `() =>` was used to capture mutable state. I already made that mistake and it took some time to find culprit. 

> The reason I ask is because you already dit this for tables in here. So its odd that you did one but not the others.

Well, doing it for tables made me realize how much such stuff lives in codebase. I saw those 2 classes, but probably there are more. So if I did it for Tables, Indexer then why not for others? ;) This PR is already big and I strongly believe in small (if possible), with well-defined scope, short-living ones. That's why I'm pushing all of those things into smaller, auxiliary tasks. They're relatively easy so maybe even someone else will do them, who knows :D

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:153]:
```suggestion
              new IllegalStateException("Language client wasn't connected!")
```
?

File: metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/metals/MetalsLanguageServer.scala:263]:
```suggestion
    case ServerState.Initialized(service) => service
```

File: metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala:24]: So this not implements all possible methods from LSP spec? Even if it isn't yet implemented by Metals?

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:37]: There are only the ones implemented by Metals?

So now when adding a new functionality, you would need to add the method in 3 places? Could you add a comment about it in each of the classes we might need to add it, as noticable as possible? :sweat_smile: 

There was already a bit of issues with the DelegatingMetalsClient etc. (that is still needed, no?)

**Review by tgodzik — CHANGES_REQUESTED** [review]
This is a very epic and courageous PR ! Very well done! I added some comments, mostly minor stuff and you will probably need to rebase on the semantic tokens PR, but that should really be just one method to add.

File: metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala:24]: No, it contains only those which are implemented by Metals

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:847]: you're right, I can't recall why I did that

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:840]: definitely, 95% of this method is just setting LSP capabilities

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala:2681]: yeah, it's leftover

File: metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/metals/lsp/DelegatingScalaService.scala:37]: Changed in [dff67d4](https://github.com/scalameta/metals/pull/4776/commits/dff67d4fb2d241a03bf340a62114eb44231a0a8c). Now I pass null as initial value to delegating service (still fine as it's not called until initialize and correct value is set before). I got rid of default implementations so now adding new endpoint in trait will cause compilation errors in both DelegatingService and in MetalsLspService (which are exactly 2 places which needs to be changed)

**Review by ckipp01 — APPROVED** [review]

**Review by tgodzik — APPROVED** [review]
LGTM! This looks great! Go ahead with the merge :D

**Pull request merged**
